api/src/org/labkey/api/assay/AbstractAssayProvider.java

```diff
@@ -1202,8 +1206,13 @@ public static void addInputMaterials(ExpRun expRun, User user, Map<ExpMaterial,
                         ExpSampleType st = newInput.getSampleType();
                         role = st != null ? st.getName() : "Sample";
                     }
-                    protApp.addMaterialInput(user, newInput, role);
+
+                    inputGroups.putIfAbsent(role, new HashSet<>());
+                    inputGroups.get(role).add(newInput.getRowId());

```

Comment:
`computeIfAbsent()` is a little more convenient and doesn't create HashSets that are ignored. Untested code:

```suggestion
                    inputGroups.computeIfAbsent(role, r -> new HashSet<>()).add(newInput.getRowId());
```